lib/dry/container/item.rb

```diff
@@ -4,22 +4,31 @@ class Container
     #
     # @private
     class Item
-      attr_reader :item, :options
+      attr_reader :item, :options, :lazy_mutex, :lazy_item
 
       def initialize(item, options = {})
         @item = item
         @options = {
           call: item.is_a?(::Proc) && item.parameters.empty?
         }.merge(options)
+        @lazy_mutex = Mutex.new if item.is_a?(::Proc) && options[:lazy] == true
       end
 
       def call
+        return lazy_item if options[:lazy] == true
```

Comment:
It would be better to set `@lazy = options[:lazy]` in the constructor, add `attr_reader :lazy` and here just do `if lazy` this way we'd avoid hash lookup for each call